tripy/tripy/frontend/utils.py

```diff
@@ -245,26 +245,39 @@ def wrapper(*args, **kwargs):
 
             all_args = utils.merge_function_arguments(func, *args, **kwargs)
 
-            # TODO (#233): Disallow mixing Tensor/Shape. The workaround below works
-            # because ShapeScalars will typically be broadcasted in order to operate
-            # with Tensors/Shapes. Otherwise, Shape and ShapeScalar would have been
-            # at the same level of hierarchy.
-            #
-            # When a function has multiple arguments, the order of precedence is:
-            #
-            # 1. Tensor
-            # 2. Shape
-            # 3. ShapeScalar
-            #
-            # That is, if we have an operation with a mixture of types, all arguments are
-            # converted to the type with the highest precedence, e.g. Tensor + Shape -> Tensor.
-            TensorType = None
-            MaxType = max(
-                (type(arg) for arg_name, arg in all_args if arg_name not in exclude),
-                key=lambda typ: {Tensor: 2, Shape: 1, ShapeScalar: 0}.get(typ, -1),
+            # Disallow mixing Tensor and Shape by default. If it makes sense in a given function
+            # to have both Tensor and Shape arguments, that might suggest that custom handling
+            # rather than relying on this decorator would make sense.
+            types = set(
+                {
+                    # There are other subclasses of Tensor, like Parameter and DefaultParameter.
+                    # Unless otherwise specified, we treat them as ordinary Tensors.
+                    Tensor if type(arg) != Shape and type(arg) != ShapeScalar else type(arg)
+                    for arg_name, arg in all_args
+                    if isinstance(arg, Tensor) and arg_name not in exclude
+                }
             )
-            if issubclass(MaxType, Tensor):
-                TensorType = MaxType
+            # We usually can treat ShapeScalars as either tensors or shapes due to broadcasting, so we can remove them from the below check.
+            shape_scalar_encountered = ShapeScalar in types
+            if ShapeScalar in types:
+                types.remove(ShapeScalar)
```

Comment:
This is equivalent to:
```suggestion
            types -= {ShapeScalar}
```

Reply:
Good to know, I forgot to wrap the RHS in a set and was wondering why it didn't work.